test-interface/src/main/scala/org/scalajs/testinterface/HTMLRunner.scala

```diff
@@ -20,130 +20,239 @@ import sbt.testing._
 protected[testinterface] object HTMLRunner extends js.JSApp {
   private val classLoader = new ScalaJSClassLoader(js.Dynamic.global)
 
-  private val logger = new Logger {
-    val ansiCodesSupported = false
-    def error(msg: String): Unit = UI.log(msg, "error")
-    def warn(msg: String): Unit = UI.log(msg, "warn")
-    def info(msg: String): Unit = UI.log(msg, "info")
-    def debug(msg: String): Unit = UI.log(msg, "debug")
-    def trace(t: Throwable): Unit = UI.log(t.toString, "error")
-  }
+  private object EventCounter {
+    private val isErrorStatus = Set(Status.Error, Status.Failure)
 
-  private object EventCounter extends EventHandler {
     val counts = mutable.Map.empty[Status, Int].withDefaultValue(0)
-    def handle(event: Event): Unit = counts(event.status) += 1
+
+    class Handler extends EventHandler {
+      private[this] var _hasErrors = false
+
+      def handle(event: Event): Unit = {
+        val status = event.status
+        _hasErrors ||= isErrorStatus(status)
+        counts(status) += 1
+      }
+
+      def hasErrors: Boolean = _hasErrors
+    }
   }
 
   def main(): Unit = {
-    val runResults = for {
-      (framework, taskDefs) <- TestDetector.detectTests()
-    } yield {
-      logger.debug(s"Detected ${taskDefs.size} tests for ${framework.name}")
-      val result = runTests(framework, taskDefs)
+    val allTests = TestDetector.detectTests()
 
-      // Report results.
-      result.filter(_.nonEmpty).foreach(r =>
-          logger.info(s"${framework.name()} reported $r"))
+    val ui = new UI(testCount = allTests.map(_._2.size).sum)
 
-      result.map(_ => ())
+    val oks = for {
+      (framework, taskDefs) <- allTests
+    } yield {
+      runTests(framework, taskDefs, ui)
     }
 
     // Report event counts.
-    Future.sequence(runResults).foreach { (_: Seq[Unit]) =>
-      import EventCounter.counts
+    Future.sequence(oks).map(and).foreach(ui.done)
```

Comment:
Ah... Indeed. I'd rather fix this thoroughly in a separate PR though. Filed as #2480.